### PR TITLE
Feat/403 create endpoint to validate location codes

### LINKED ISSUE #403: Create endpoint to validate location codes
In order to validate the location codes that users will provide, we need a new endpoint to fetch the full location from the Forest Client API. Providing the client number and the location code that it's going to be validated, the API will either return the full location, if the location code is valid, or an message saying that there are no locations for that code.

Requirements:

Create endpoints to validate client's location code from the [FSA Forest Client API ](https://nr-forest-client-api-prod-backend.apps.silver.devops.gov.bc.ca/v3/swagger-ui/index.html).

/api/forest-clients/{clientNumber}/location/{location}​: fetch a location given its identifier and its client's number

The application must call the appropriate Forest Client API endpoint and return the response to the client.

### PR DESCRIPTION
**Pull request by mgaseta**
# Description

Creating a new endpoint that returns a client location based on the client number and location code.
This will be used to validate location codes provided by the user on the necessary forms of the frontend. 

Fixes #403 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Manually tested via swagger and ran unit tests.

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have already been accepted and merged

## Further comments

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend](https://nr-spar-410-backend.apps.silver.devops.gov.bc.ca/)
[Frontend](https://nr-spar-410-frontend.apps.silver.devops.gov.bc.ca/)
[Oracle-API](https://nr-spar-410-oracle-api.apps.silver.devops.gov.bc.ca/)

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge-main.yml)

### CODE REVIEW

**Review by craigyu — APPROVED** [review]
Nicely done! 
image

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java [hunk 1]
```diff
@@ -105,4 +105,42 @@ public List<ForestClientLocationDto> fetchClientLocations(
...
+              Returns a single location associated with the forest client, identified
+              by it's number and location code""",
+      responses = {
+        @ApiResponse(
+            responseCode = "200"),
+        @ApiResponse(responseCode = "404", content = @Content(schema = @Schema(hidden = true)))
+      })
+  public ForestClientLocationDto fetchSingleClientLocation(
+      @PathVariable("number")
+          @Pattern(regexp = "^\\d{8}$", message = "The value must an 8-digit number")
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java:131]: I think you missed **be** in **The value must...**

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java [hunk 2]
```diff
@@ -105,4 +105,42 @@ public List<ForestClientLocationDto> fetchClientLocations(
...
+          """
+              Returns a single location associated with the forest client, identified
+              by it's number and location code""",
+      responses = {
+        @ApiResponse(
+            responseCode = "200"),
+        @ApiResponse(responseCode = "404", content = @Content(schema = @Schema(hidden = true)))
+      })
+  public ForestClientLocationDto fetchSingleClientLocation(
+      @PathVariable("number")
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java:130]: Maybe we could use a meaningful name, like `clientNumber` ?

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java [hunk 3]
```diff
@@ -105,4 +105,42 @@ public List<ForestClientLocationDto> fetchClientLocations(
           String number) {
     return forestClientService.fetchClientLocations(number);
   }
+
+  /**
+   * Fetch the forest client location based on the client number and location code.
+   *
+   * @param number the number that identifies the client to fetch the location
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java:112]: I think there's a preposition thing here 🤣 when I read, it sounds it's missing **from** in the end. [to fetch the location *from*]

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java [hunk 4]
```diff
@@ -105,4 +105,42 @@ public List<ForestClientLocationDto> fetchClientLocations(
...
+   * @param locationCode the location code that identifies the location to be fetched
+   * @return the forest client location
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java:114]: We could make a better use of JavaDoc, pointing out what will be returned (class). Like `@return {@link ForestClientLocationDto} containing client and location data.`

File: backend/src/main/java/ca/bc/gov/backendstartapi/provider/ForestClientApiProvider.java [hunk 5]
```diff
@@ -138,6 +139,38 @@ public List<ForestClientLocationDto> fetchLocationsByClientNumber(String number)
     return List.of();
   }
 
+  /**
+   * Fetch a single location of a forest client its number and location code.
+   *
+   * @param number the client number to search the location
+   * @param locationCode the location code that identifies the location to be fetched
+   * @return a single location for a forest client
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/provider/ForestClientApiProvider.java:147]: Same here. Improve docs 😄

File: backend/src/test/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpointTest.java [hunk 6]
```diff
@@ -206,4 +208,78 @@ void fetchNonExistentClientLocations() throws Exception {
...
+    when(forestClientService.fetchSingleClientLocation("00000000", "12"))
+        .thenThrow(new ResponseStatusException(HttpStatus.NOT_FOUND));
+
+    mockMvc
+        .perform(
+            get("/api/forest-clients/00000000/locations/12")
+                .with(csrf().asHeader())
+                .header("Content-Type", MediaType.APPLICATION_JSON_VALUE)
+                .accept(MediaType.APPLICATION_JSON))
+        .andExpect(status().isNotFound());
```
Comment (RMCampos) [inline backend/src/test/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpointTest.java:283]: 👍🏻 good job

File: backend/src/test/java/ca/bc/gov/backendstartapi/provider/ForestClientApiProviderTest.java [hunk 7]
```diff
@@ -244,4 +246,85 @@ void fetchNonExistentClientLocation() {
...
+        .andRespond(withStatus(HttpStatusCode.valueOf(404)));
+
+    ResponseStatusException exc =
+        Assertions.assertThrows(
+          ResponseStatusException.class,
+          () -> {
+            forestClientApiProvider.fetchSingleClientLocation(number, locationCode);
+          });
+
+    Assertions.assertEquals(HttpStatus.NOT_FOUND, exc.getStatusCode());
```
Comment (RMCampos) [inline backend/src/test/java/ca/bc/gov/backendstartapi/provider/ForestClientApiProviderTest.java:328]: 👍🏻 awesome

File: backend/src/test/java/ca/bc/gov/backendstartapi/service/ForestClientServiceTest.java [hunk 8]
```diff
@@ -146,4 +148,74 @@ void fetchClientLocationSuccess() {
...
+    Assertions.assertEquals("ARROW FOREST DISTRICT", locationDto.address2());
+    Assertions.assertEquals("845 COLUMBIA AVENUE", locationDto.address3());
+    Assertions.assertEquals("CASTLEGAR", locationDto.city());
+    Assertions.assertEquals("BC", locationDto.province());
+    Assertions.assertEquals("V1N1H3", locationDto.postalCode());
+    Assertions.assertEquals("CANADA", locationDto.country());
+    Assertions.assertEquals("2503658600", locationDto.businessPhone());
+    Assertions.assertEquals(ForestClientExpiredEnum.N,
+                            ForestClientExpiredEnum.valueOf(locationDto.expired().name()));
+    Assertions.assertEquals(ForestClientExpiredEnum.N,
```
Comment (RMCampos) [inline backend/src/test/java/ca/bc/gov/backendstartapi/service/ForestClientServiceTest.java:218]: 👍🏻

**Review by RMCampos — CHANGES_REQUESTED** [review]
Great job. Only minor changes related to JavaDocs. 💯 🥇 

Also, how about updating all references to **forest client** as **ForestClient**? I think this can make it more clear, that's another products.

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java (hunk 1 above)
Comment (mgaseta) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java]: Clinical eye :detective: :eyes: 
Fixed https://github.com/bcgov/nr-spar/pull/410/commits/59e43eb7d69f503b0c65962d1a34e6732a4cbcf0

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java (hunk 3 above)
Comment (mgaseta) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java]: Re-wrote the whole sentence, it makes more sense now (I think hahahah :stuck_out_tongue:)
Fix here https://github.com/bcgov/nr-spar/pull/410/commits/59e43eb7d69f503b0c65962d1a34e6732a4cbcf0

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java (hunk 4 above)
Comment (mgaseta) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java]: Much better this way, thanks!
Fix here: https://github.com/bcgov/nr-spar/pull/410/commits/59e43eb7d69f503b0c65962d1a34e6732a4cbcf0

File: backend/src/main/java/ca/bc/gov/backendstartapi/provider/ForestClientApiProvider.java (hunk 5 above)
Comment (mgaseta) [reply backend/src/main/java/ca/bc/gov/backendstartapi/provider/ForestClientApiProvider.java]: Aye aye.
Fixed here: https://github.com/bcgov/nr-spar/pull/410/commits/59e43eb7d69f503b0c65962d1a34e6732a4cbcf0

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java (hunk 2 above)
Comment (mgaseta) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java]: Makes sense.
Fixed here: https://github.com/bcgov/nr-spar/pull/410/commits/350d7cc863e369071e6266ae974969fad7d33350

**Review by RMCampos — APPROVED** [review]
Great job!! 🥇 💯 🎉

**Pull request merged**